beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -651,6 +649,31 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
         Ok(iter.map(|result| result.map_err(Into::into)))
     }
 
+    /// Super-efficient forwards state roots iterator that avoids cloning the head if the state
+    /// roots lie entirely within the freezer database.
+    ///
+    /// The iterator returned will include roots for `start_slot..=end_slot`, i.e.  it
+    /// is endpoint inclusive.
+    pub fn forwards_iter_state_roots_until(
+        &self,
+        start_slot: Slot,
+        end_slot: Slot,
+    ) -> Result<impl Iterator<Item = Result<(Hash256, Slot), Error>> + '_, Error> {
+        self.with_head(move |head| {
```

Comment:
Ah, good catch.

It's ok in the case where the iteration starts in the frozen root section of the DB, because we clone the head and return the iterator (which is lazy) here:

https://github.com/sigp/lighthouse/blob/ec830f1f77d8aa92cf616a06c1b10ae39f9886e6/beacon_node/store/src/forwards_iter.rs#L285-L294

However in the case where we start after the `last_restore_point_slot` we eagerly iterate backwards from the head while holding the lock, which is not ideal:

https://github.com/sigp/lighthouse/blob/ec830f1f77d8aa92cf616a06c1b10ae39f9886e6/beacon_node/store/src/forwards_iter.rs#L296-L304

I'll add a variant to the hybrid iterator that represents the "between state" of just holding the state before we start iterating backwards. Then we can preserve the property that the iterator is always quick to construct

Reply:
I've implemented this in https://github.com/sigp/lighthouse/pull/2863/commits/b892bb79f7c73b4a93b73814b95bf344f10ca486 and ended up consolidating the block root and state root iterators while I was there.